rules/prevent-abbreviations.js

```diff
@@ -0,0 +1,369 @@
+'use strict';
+const defaultsDeep = require('lodash.defaultsdeep');
+const toPairs = require('lodash.topairs');
+
+const getDocsUrl = require('./utils/get-docs-url');
+const avoidCapture = require('./utils/avoid-capture');
+
+const defaultReplacements = {
+	err: {
+		error: true
+	},
+	cb: {
+		callback: true
+	},
+	opts: {
+		options: true
+	},
+	str: {
+		string: true
+	},
+	obj: {
+		object: true
+	},
+	num: {
+		number: true
+	},
+	val: {
+		value: true
+	},
+	e: {
+		event: true,
+		error: true
+	},
+	evt: {
+		event: true
+	},
+	el: {
+		element: true
+	},
+	req: {
+		request: true
+	},
+	res: {
+		response: true,
+		result: true
+	},
+	btn: {
+		button: true
+	},
+	msg: {
+		message: true
+	},
+	len: {
+		length: true
+	},
+	env: {
+		environment: true
+	},
+	dev: {
+		development: true
+	},
+	prod: {
+		production: true
+	},
+	tmp: {
+		temporary: true
+	},
+	arg: {
+		argument: true
+	},
+	args: {
+		arguments: true
+	},
+	tbl: {
+		table: true
+	},
+	db: {
+		database: true
+	},
+	ctx: {
+		context: true
+	},
+	mod: {
+		module: true
+	},
+	prop: {
+		property: true
+	}
+};
+
+const prepareOptions = ({extendDefaults = true, replacements = {}} = {}) => {
+	const mergedReplacements = extendDefaults ?
+		defaultsDeep({}, replacements, defaultReplacements) :
+		replacements;
+
+	return new Map(toPairs(mergedReplacements).map(([discouragedName, replacements]) => {
+		return [discouragedName, new Map(toPairs(replacements))];
+	}));
+};
+
+const getVariableReplacements = (replacements, variable) => {
+	const variableNameReplacements = replacements.get(variable.name);
+
+	if (!variableNameReplacements) {
+		return [];
+	}
+
+	return [...variableNameReplacements.keys()]
+		.filter(name => variableNameReplacements.get(name))
+		.sort();
+};
+
+const scopeHasArgumentsSpecial = scope => {
+	while (scope) {
+		if (scope.taints.get('arguments')) {
+			return true;
+		}
+
+		scope = scope.upper;
+	}
+
+	return false;
+};
+
+const collideWithArgumentsSpecial = (names, scopes) => {
+	if (!names.includes('arguments')) {
+		return false;
+	}
+
+	return scopes.some(scopeHasArgumentsSpecial) ||
+		scopes.some(scope => scope.isStrict);
+};
+
+const anotherNameMessage = 'A more descriptive name will do too.';
+
+const formatMessage = (discouragedName, replacements) => {
+	const message = [];
+
+	if (replacements.length === 1) {
+		message.push(`This variable should be named \`${replacements[0]}\`.`);
```

Comment:
I think we should mention the original variable name to make it faster to understand what needs changing.

How about this?

```
The variable `evt` should be named `event`.
```